### PR TITLE
imxrt: fixed CAN error while sending larger amount of data

### PR DESCRIPTION
**Pull request by michallenc**
## Summary
FlexCAN on imxrt was loosing some messages while sending larger amount of packages (tousands or ten tousands). This was caused by Error and Status Register 2 holding older value even if some MB was free – added function that waits for register to update and then sends data.

## Testing

Tested on Teensy 4.1 board and imxrt1060-EVK using @PetervdPerk-NXP program [canpkt](https://github.com/PetervdPerk-NXP/nuttx-apps-testing/tree/master/examples/canpkt)

### CODE REVIEW

File: arch/arm/src/imxrt/imxrt_flexcan.c [hunk 1]
```diff
@@ -1183,6 +1198,27 @@ static uint32_t imxrt_waitmcr_change(uint32_t base, uint32_t mask,
   return false;
 }
 
+static uint32_t imxrt_waitesr2_change(uint32_t base, uint32_t mask,
+                                       uint32_t target_state)
```
Comment (acassis) [inline arch/arm/src/imxrt/imxrt_flexcan.c:1202]: Hi @michallenc, since the "state" is bool shouldn't it be better if target_state also be bool ?

**Comment by PetervdPerk-NXP** [comment]
Hey @michallenc thank you for you contribution, by finding this issue with the ESR2 register, however when stress testing the CAN TX still dropped frames will occur. I've made an alternative fix, could you test that one?

**Comment by michallenc** [comment]
> Hey @michallenc thank you for you contribution, by finding this issue with the ESR2 register, however when stress testing the CAN TX still dropped frames will occur. I've made an alternative fix, could you test that one?

Hi @PetervdPerk-NXP, thank you for checking, your changes are working on my EVK, however it seems the received packages are mixed up:

```
  can4  123   [8]  50 00 00 00 00 00 00 00
  can4  123   [8]  4F 00 00 00 00 00 00 00
  can4  123   [8]  52 00 00 00 00 00 00 00
  can4  123   [8]  51 00 00 00 00 00 00 00
  can4  123   [8]  54 00 00 00 00 00 00 00
  can4  123   [8]  53 00 00 00 00 00 00 00
  can4  123   [8]  56 00 00 00 00 00 00 00
  can4  123   [8]  55 00 00 00 00 00 00 00
  can4  123   [8]  58 00 00 00 00 00 00 00
  can4  123   [8]  57 00 00 00 00 00 00 00
  can4  123   [8]  5A 00 00 00 00 00 00 00
  can4  123   [8]  59 00 00 00 00 00 00 00
  can4  123   [8]  5C 00 00 00 00 00 00 00
  can4  123   [8]  5B 00 00 00 00 00 00 00
  can4  123   [8]  5E 00 00 00 00 00 00 00
```

Replacing if (!imxrt_txringfull(priv)) in function imxrt_txavail_work on line 1297 with registers the same way you did in imxrt_transmit seems to address this problem.

```
  can4  123   [8]  51 00 00 00 00 00 00 00
  can4  123   [8]  52 00 00 00 00 00 00 00
  can4  123   [8]  53 00 00 00 00 00 00 00
  can4  123   [8]  54 00 00 00 00 00 00 00
  can4  123   [8]  55 00 00 00 00 00 00 00
  can4  123   [8]  56 00 00 00 00 00 00 00
  can4  123   [8]  57 00 00 00 00 00 00 00
  can4  123   [8]  58 00 00 00 00 00 00 00
  can4  123   [8]  59 00 00 00 00 00 00 00
  can4  123   [8]  5A 00 00 00 00 00 00 00
  can4  123   [8]  5B 00 00 00 00 00 00 00
  can4  123   [8]  5C 00 00 00 00 00 00 00
  can4  123   [8]  5D 00 00 00 00 00 00 00
  can4  123   [8]  5E 00 00 00 00 00 00 00
  can4  123   [8]  5F 00 00 00 00 00 00 00
```

Would you test that change in your code? If it does work with your board I will close this PR so you can push yours.

**Comment by PetervdPerk-NXP** [comment]
Replacing if (!imxrt_txringfull(priv)) in function imxrt_txavail_work on line 1297 solves the ordering issue indeed. However now sometimes it causes that canpkt -t runs into deadlock, I've got to make some free time this week to debug this further. Thanks for the help and feedback though.

**Comment by michallenc** [comment]
> Replacing if (!imxrt_txringfull(priv)) in function imxrt_txavail_work on line 1297 solves the ordering issue indeed. However now sometimes it causes that canpkt -t runs into deadlock, I've got to make some free time this week to debug this further. Thanks for the help and feedback though.

Thanks. I am closing the PR as there is no need for my changes now.

**Pull request closed without merging**